### ISSUE TITLE
Proposed rewording of section 12.5.  Unlinkability

### ISSUE THREAD

**Issue by Denisthemalice** [oauth-wg/draft-ietf-oauth-status-list#207]
It is proposed to include another section called : 

> 12.5.2  Alternative mechanism to the use of Token Status Lists

Before it, a section that will re-use the existing text will be called :

> 12.5.1  Further recommendations to the use of Token Status Lists

The whole text would look as follows:

**12.5.1  Further recommendations to the use of Token Status Lists**

   To avoid privacy risks for colluding Relying Parties, it is RECOMMENDED 
   that Issuers use batch issuance to issue multiple tokens, see Section 13.1.

   To avoid further correlatable information by the values of uri and
   index, Issuers are RECOMMENDED to:

   *  choose non-sequential, pseudo-random or random indices

   *  use decoy entries to obfuscate the real number of Referenced
      Tokens within a Status List

   *  choose to deploy and utilize multiple Status Lists simultaneously

**12.5.2  Alternative mechanism to the use of Token Status Lists**

   To avoid privacy risks for colluding Relying Parties, Issuers SHOULD 
   use batch issuance of multiple Tokens, but it is possible to use a 
   mechanism different from the use of Token Status Lists.   In such a 
   case, the cryptographically secured data structure derived from the 
   Token by the Holder does not contain a reference to a Token Status 
   List. 
      
   This mechanism is only usable if the Issuer has been able to verify, 
   at the time of the issuance of a Token, that an appropriate application
   running under a secure environment is supported by the Holder.

   Instead of using a protocol for fetching token status lists from a 
   Status Issuer, it is possible to take advantage of another protocol 
   between an Holder and an Issuer, when that protocol is supported.

   The fact that this other protocol is supported by an appropriate 
   application running under a secure environment can be indicated by 
   the Issuer to the Verifier by including into the Token, for example, 
   a security policy similar to the certificate policy from X.509 certificates 
   (RFC 5280). 
   See: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.4

   When the Relying Party is able to know that this protocol between 
   an Holder and an Issuer is supported, it knows that the Issuer will 
   be able, when necessary, to either suspend or invalidate any Token 
   that is has issued. 
    
   The protocol works as follows: 

   from time to time, the Holder queries each Issuer that has issued 
   a Token, and if a Token that it has issued needs to be suspended 
   or invalidated, the Issuer informs the Holder.   In that case, the 
   Holder will  immediately stop creating new cryptographically 
   secured data structures derived from suspended or invalidated 
   Tokens.

   When a Token has been invalidated, the process cannot be reversed. 
   However, when a Token has been suspended, the process can be reversed.

   If the Holder has been unable to contact an Issuer during a defined 
   period of time, e.g., 24 hours, the Holder will stop creating new 
   cryptographically secured data structures derived from Tokens issued
   by that Issuer.

   In this way, as long as the Holder is able to connect to each of the 
   Issuers from time to time, this mechanism can be used in an off-line 
   environment. 

   In order to check the suspension or the revocation status of a 
   cryptographically secured data structure, neither the Holder nor the 
   Verifier needs to fetch additional information.  When digital 
   credentials are used, the Holder and the Verifier can interact in a 
   proximity mode without using any on-line protocol.

**Comment by paulbastian** [comment]
This draft not only introduces the Status List mechanism but also an extension point, that may be used to introduce other status mechanisms and establishes a registry. If you want to pursue your idea, I would recommend to make a new draft for this, as it seems too different from the Status List mechanism, then add an IANA consideration section that registers it as a new status mechanism.

**Comment by Denisthemalice** [comment]
At the moment, the reader might believe that there exits is no alternative method supporting the verifier-verifier unlinkability property and if this property is desirable this makes the design particularly difficult.

I read section 12.5 and I have not been convinced that the RECOMMENDATIONS placed in this section to avoid further correlatable information by the values of uri and index, can fully address the unlinkability issue.

The key point is that I would like to advertise to the readers that there exists an alternative method to the "status_list" method that is able to support very easily the verifier-verifier unlinkability property.

I read your comment and you are right: this new method could be added to JWT Status Mechanism Methods Registry.

The use of the status_list method requires the use of a protocol to fetch the status list. However, there is no magic: if this protocol is not used, another protocol shall be used. This protocol shall take place between the Holder and the Provider of the Holder (application). 

Such protocols between Holders and the Providers of the Holder already exist, but are proprietary (e.g. to steal the contacts of the individuals or to propose app. updates). As it would be unrealistic to define an interoperable protocol, this cannot end-up with a standard track RFC, but this method could be described in an Informational RFC. 

A problem is that this method is only one piece of a puzzle of a larger picture that would describe the architecture of the three roles Model and such a document is not currently on the agenda of this WG. In particular, the description of this architecture would start with an inital request of a "Token" to an Issuer that must be done by a Holder running under a secure environment. 

I would be happy to contribute to it, but I cannot take the lead on working draft document as my knowledge of the use of github is far from being sufficient.

Instead of my initial text proposal, I propose the following text which has been slightly revised :

> To avoid privacy risks for colluding Relying Parties, Issuers SHOULD use batch issuance of multiple Tokens, but it is possible to use a method different from the use of Token Status Lists. In such a case, the cryptographically secured data structure derived from the Token by the Holder would contain a Status Method Value different  from "status_list".
> 
> This method is only usable if the Issuer has been able to verify, at the time of the issuance of a Token, that an appropriate application running under a secure environment is supported by the Holder.
> 
> Instead of using a protocol for fetching token status lists from a Status Issuer, it is possible to take advantage of another protocol between an Holder and an Issuer.
> 
> When the Relying Party knows that this other method is supported, it knows that the Issuer will be able, when necessary, to either suspend or invalidate any Token that is has issued.
> 
> The protocol works as follows:
> 
> From time to time, the Holder queries each Issuer that has issued a Token, and if a Token that it has issued needs to be suspended or invalidated, the Issuer informs the Holder. In that case, the Holder will immediately stop creating new cryptographically secured data structures derived from suspended or invalidated Tokens.
> 
> When a Token has been invalidated, the process cannot be reversed.
> However, when a Token has been suspended, the process can be reversed.
> 
> If the Holder has been unable to contact an Issuer during a defined period of time, e.g., 24 hours, the Holder will stop creating new cryptographically secured data structures derived from Tokens issued by that Issuer.
> 
> In this way, as long as the Holder is able to connect to each of the Issuers from time to time, this mechanism can be used in an off-line environment.
> 
> In order to check the suspension or the revocation status of a cryptographically secured data structure, neither the Holder nor the Verifier needs to fetch additional information. The Holder and the Verifier can interact in a proximity mode without using any on-line protocol.
> 
>

**Comment by paulbastian** [comment]
If it is not clear enough that the privacy and security considerations only apply to the status list, then we can improve this. Butt we should not confuse the reader therefore I think that one draft should only describe one status mechanism.

**Comment by Denisthemalice** [comment]
I will reuse my previous arguments but shuffling them a bit:

I read section 12.5 and I have not been convinced at all that the RECOMMENDATIONS placed in this section allow to avoid further correlatable information by the values of uri and index.

IMO, the ability to support the verifier-verifier unlinkability property is VERY important. 
It should not be ruined by the use of Token status lists.

At the moment, the reader might believe that there exits no method really supporting the verifier-verifier unlinkability property. 
I would like to inform the readers of this document that there exists an alternative method to the "status_list" method that is able to support (very easily) the verifier-verifier unlinkability property.

This method would greatly simplify the code that the Relying Party /  Verifier will need to support.

The core of the document will continue to describe only one status mechanism.
The description of this alternative method would be placed into the Privacy Considerations section.

**Comment by paulbastian** [comment]
> I will reuse my previous arguments but shuffling them a bit:
> 
> I read section 12.5 and I have not been convinced at all that the RECOMMENDATIONS placed in this section allow to avoid further correlatable information by the values of uri and index.
> 
> IMO, the ability to support the verifier-verifier unlinkability property is VERY important. 
> It should not be ruined by the use of Token status lists.

I agree that verifier-verifier unlinkability is very important. Could you elaborate why you think the recommendations are insufficient to guarantee this for status list? 

> 
> At the moment, the reader might believe that there exits no method really supporting the verifier-verifier unlinkability property. 
> I would like to inform the readers of this document that there exists an alternative method to the "status_list" method that is able to support (very easily) the verifier-verifier unlinkability property.

As I said, we can add text that explains that other status mechanisms may have other tradeoffs regarding privacy, scalability or complexity.

> 
> This method would greatly simplify the code that the Relying Party /  Verifier will need to support.

Your method simplifies things for the Relying Party but puts other burdens on holders and issuers. Revocation is complicated and there are no easy tradeoffs, we shouldn't let the reader believe this.

> 
> The core of the document will continue to describe only one status mechanism.
> The description of this alternative method would be placed into the Privacy Considerations section.

The alternative status mechanism should have a dedicated draft and within this draft the privacy consideration section may explain this.

**Comment by Denisthemalice** [comment]
Clause 12.5 (Unlinkability) states:

>    Colluding Issuers and Relying Parties have the possibility to link
>    two transactions, as the tuple of uri and index inside the Referenced
>    Token are unique and therefore traceable data.  

This is sufficient to link two transactions without the need to fetch the Token Status List. 
Whatever will be done for the management of the Token Status list, this linkage will remain. 
I probably missed something: what did I missed?

**Comment by paulbastian** [comment]
> Clause 12.5 (Unlinkability) states:
> 
> >    Colluding Issuers and Relying Parties have the possibility to link
> >    two transactions, as the tuple of uri and index inside the Referenced
> >    Token are unique and therefore traceable data.  
> 
> This is sufficient to link two transactions without the need to fetch the Token Status List. 
> Whatever will be done for the management of the Token Status list, this linkage will remain. 
> I probably missed something: what did I missed?
> 

This is correct. Token Status List can achieve verifier-verifier unlinkability but not issuer-verifier unlinkability. I think this is sufficiently described in the text, you do not agree?

**Comment by Denisthemalice** [comment]
No. I meant: Token Status List CANNOT achieve verifier-verifier unlinkability ... unless you demonstrate the contrary.

**Comment by paulbastian** [comment]
If you follow the recommendations in privacy consideration, you will achieve this. Every reference token gets it's separate status index using batch issuance and with randomized indices, how do you want to correlate?

**Comment by Denisthemalice** [comment]
You are right for the verifier-verifier unlinkability, I forgot the main point: 
if a Token is used only once (which means that batch issuance is used), there cannot be any correlation between verifiers by using the status claim.

I re-read the text and I believe that it is not crystal clear. 

The current text is as follows:

> Colluding Issuers and Relying Parties have the possibility to link two transactions, as the tuple of uri and index inside the Referenced Token are unique and therefore traceable data.  By comparing the status claims of received Referenced Tokens, two colluding Relying Parties could determine that they have interacted with the same user or an Issuer could trace the usage of its issued Referenced Token by colluding with various Relying Parties. 

> It is therefore recommended to use Status Lists for Referenced Token formats that have similar unlinkability properties.

The last sentence only applies to the Colluding Issuers and Relying Parties case.

I would recommend to separate the two cases:

> 12.5.1  Colluding Relying Parties and
> 12.5.2  Colluding Issuers and Relying Parties.

In order to clarify the first case, I would propose:

> 12.5.  Unlinkability
> 
> The tuple of uri and index inside the Referenced Token are unique and therefore is traceable data.
> 
> 12.5.1  Colluding Relying Parties
> 
> By comparing the status claims of received Referenced Tokens, two colluding Relying Parties could determine that they have interacted with the same user.  To avoid privacy risks for colluding Relying Parties, it is RECOMMENDED that Issuers use batch issuance to issue multiple tokens, see Section 13.1. In this case, a Referenced Token SHALL be presented only once to a Relying Party.

Then, I attempted to understand the second case:

> (...) or an Issuer could trace the usage of its issued Referenced Token by colluding with various Relying Parties.

This linkage can simply happen by comparing Referenced Tokens originating from an Issuer and Referenced Tokens received by a Relying Party, i.e., without using the content of the Status List.  However, the remaining of this section is addressing considerations that apply to the content of the Status Lists.

I believe that the text would need to be clarified.

Does it apply to "Status Issuers" rather than to "Issuers" ?

I would also like to make an observation. If a Status Issuer is willing to be able to trace the usage of its issued Referenced Token by colluding with various Relying Parties, it will NOT implement these RECOMMENDATIONS and nobody will know whether it has implemented them or not (at least, the first two recommendations).

In a previous response, you wrote:

> Token Status List can achieve verifier-verifier unlinkability but not issuer-verifier unlinkability.

However, while reading the current text, it looks that the three RECOMMENDATIONS allow to remove the risk. 
This is confusing.

**Comment by paulbastian** [comment]
Editors proposal:
- clarify that other status mechanisms have other privacy and security properties (and this spec defines an extension point for new status mechanisms)
- differentiate issuer-verifier and verifier-verifier unlinkability and clarify that the privacy recommendations only apply to verifier-verifier unlinkability
- change Issuer to Status Issuer for the linked text

**Comment by Denisthemalice** [comment]
Well,

Would you be able to propose a revised wording ?

I cought the following sentence that you used:

> Revocation is complicated and there are no easy tradeoffs, we shouldn't let the reader believe this

I responded to it by opening a new issue #209 :

> Revocation and suspension are not that complicated and there are easy tradeoffs

:-)
